doc/source/data/performance-tips.rst

```diff
@@ -25,17 +25,122 @@ Tuning read parallelism
+3. Max block size (default=128MiB). If the parallelism would make blocks larger than this threshold, the parallelism is increased to avoid OOMs during processing. This can be overridden by setting :class:`DataContext.target_max_block_size <ray.data.context.DataContext>` (bytes).
+4. Available CPUs. If the parallelism cannot make use of all the available CPUs in the cluster, the parallelism is increased until it can. Ray Data chooses read tasks be at least 2x the number of available CPUs.
+
+Occasionally, it's advantageous to manually tune the parallelism to optimize the application.
+For example, the following code will batch multiple files into the same read task to avoid creating blocks that are too large.
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Repeat the iris.csv file 16 times.
+    ds = ray.data.read_images(["example://iris.csv"] * 16)
+    print(ds.materialize())
+    # 2023-11-20 14:28:47,597 INFO plan.py:760 -- Using autodetected parallelism=4 for stage ReadCSV to satisfy parallelism at least twice the available number of CPUs (2).
+    # MaterializedDataset(
+    #    num_blocks=4,
+    #    num_rows=2400,
+    #    ...
+    # )
+
+But suppose that you knew that you wanted to read all 16 files in parallel.
+This could, for example, because you know that additional CPUs should get added to the cluster by the autoscaler or because you want the downstream stage to transform each file's contents in parallel.
+You can get this behavior by setting the ``parallelism`` parameter.
+Notice how the number of output blocks is equal to ``parallelism`` in the following code:
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Repeat the iris.csv file 16 times.
+    ds = ray.data.read_images(["example://iris.csv"] * 16, parallelism=16)
+    print(ds.materialize())
+    # MaterializedDataset(
+    #    num_blocks=16,
+    #    num_rows=2400,
+    #    ...
+    # )
+
+
+When using the default autodetected ``parallelism``, Ray Data attempts to cap each task's output to :class:`DataContext.target_max_block_size <ray.data.context.DataContext>` many bytes.
+Note however that Ray Data cannot perfectly predict the size of each task's output, so it is possible that each task produces one or more output blocks.
+Thus, the total blocks in the final :class:`~ray.data.Dataset` may differ from the specified ``parallelism``.
+Here's an example where we manually specify ``parallelism=1``, but the one task will still produce multiple blocks in the materialized Dataset:
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Generate ~400MB of data.
+    ds = ray.data.range_tensor(5_000, shape=(10_000, ), parallelism=1)
+    print(ds.materialize())
+    # MaterializedDataset(
```

Comment:
Move to `testoutput`? I don't think we'd need to do `:options: +MOCK` for this either?